src/librustc/ty/layout.rs

```diff
@@ -706,13 +707,34 @@ impl<'a, 'tcx> LayoutCx<'tcx, TyCtxt<'a, 'tcx, 'tcx>> {
                         } else {
                             align = align.max(field.align);
                         }
+
+                        if field.abi != Abi::Uninhabited {
+                            // Normalize scalar_unit to the maximal valid range
+                            let field_abi = match &field.abi {
+                                Abi::Scalar(x) => Abi::Scalar(scalar_unit(x.value)),
+                                Abi::ScalarPair(x, y) => {
+                                    Abi::ScalarPair(
+                                        scalar_unit(x.value),
+                                        scalar_unit(y.value),
+                                    )
+                                }
+                                _ => field.abi.clone(),
+                            };
+
+                            if abi == Abi::Uninhabited || size == Size::ZERO {
+                                abi = field_abi;
+                            } else if field.size != Size::ZERO && abi != field_abi  {
```

Comment:
`!field.is_zst()` is a better test because it takes DST into account. I know we don't allow these in unions, but still, no reason not to use that.